File: Model/quadrilatero.cpp
#include "quadrilatero.h"
#include <QVector>
#include <math.h>

Quadrilatero::Quadrilatero(const Punto& p1, const Punto& p2, const Punto& p3, const Punto& p4) : LatiFiniti() {
    pushVertice(p1);
    pushVertice(p2);
    pushVertice(p3);
    pushVertice(p4);
}

double Quadrilatero::getArea() const {
    QVector<double> l = getLati();
    double diag1 = Punto::distanzaDuePunti((getVertice(0)),(getVertice(2)));
    double diag2 = Punto::distanzaDuePunti((getVertice(1)),(getVertice(3)));
    double diag = std::min(diag1, diag2);

    if(diag1 <= diag2) {
        double semiP1 = (l[0]+l[1]+diag)/2;
        double area1 = sqrt(semiP1 * (semiP1-l[0]) * (semiP1-l[1]) * (semiP1-diag));
        double semiP2 = (diag+l[2]+l[3])/2;
        double area2 = sqrt(semiP2 * (semiP2-diag) * (semiP2-l[2]) * (semiP2-l[3]));
        return area1+area2;
    }
    else {
        double semiP1 = (l[0]+l[3]+diag)/2;
        double area1 = sqrt(semiP1 * (semiP1-l[0]) * (semiP1-l[3]) * (semiP1-diag));
        double semiP2 = (diag+l[1]+l[2])/2;
        double area2 = sqrt(semiP2 * (semiP2-diag) * (semiP2-l[1]) * (semiP2-l[2]));
        return area1+area2;
    }
}

Quadrilatero* Quadrilatero::clone() const {
    return new Quadrilatero(*this);
}

File: SuperQtKalkJAVA/src/Quadrilatero.java
import java.util.ArrayList;
import static java.lang.Math.sqrt;

public class Quadrilatero extends LatiFiniti {

    public Quadrilatero(Punto a, Punto b, Punto c, Punto d) {
        super(new ArrayList<Punto>());
        pushVertice(a);
        pushVertice(b);
        pushVertice(c);
        pushVertice(d);
    }
    
    public double getArea() {
        ArrayList<Double> l = getLati();
        double diag1 = Punto.distanzaDuePunti((getVertice(0)),(getVertice(2)));
        double diag2 = Punto.distanzaDuePunti((getVertice(0)),(getVertice(2)));
        double diag = Math.min(diag1, diag2);
        if(diag1 <= diag2) {
            double semiP1 = (l.get(0)+l.get(1)+diag)/2;
            double area1 = sqrt(semiP1 * (semiP1-l.get(0)) * (semiP1-l.get(1)) * (semiP1-diag));
            double semiP2 = (diag+l.get(2)+l.get(3)/2);
            double area2 = sqrt(semiP2 * (semiP2-diag) * (semiP2-l.get(2)) * (semiP2-l.get(3)));
            return area1+area2;
        }
        else {
            double semiP1 = (l.get(0)+l.get(3)+diag)/2;
            double area1 = sqrt(semiP1 * (semiP1-l.get(0)) * (semiP1-l.get(3)) * (semiP1-diag));
            double semiP2 = (diag+l.get(1)+l.get(2)/2);
            double area2 = sqrt(semiP2 * (semiP2-diag) * (semiP2-l.get(2)) * (semiP2-l.get(3)));
            return area1+area2;
        }
    }

}
